src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java

```diff
@@ -251,6 +252,30 @@ private void checkCommErrors() throws ServerException {
 		}
 	}
 
+	/**
+	 * Returns true if connected with a recent event.
+	 * If binlog heartbeats are disabled, just returns
+	 * whether there is a connection.
+	 */
+	private boolean isConnectionAlive() {
+		if (!isConnected) {
+			return false;
+		}
+		boolean isAlive = true;
+		if (client.getHeartbeatInterval() > 0L) {
+			long lastEventSeenAt = binlogEventListener.getLastEventSeenAt();
+			long lastEventAge = System.currentTimeMillis() - lastEventSeenAt;
+			long maxAllowedEventAge = Math.round(binlogHeartbeatInterval * binlogHeartbeatIntervalAllowance);
+			isAlive = lastEventAge <= maxAllowedEventAge;
+			if (!isAlive) {
+				LOGGER.warn(
+					"Last binlog event seen " + lastEventAge + "ms ago, exceeding " + maxAllowedEventAge + "ms allowance " +
+					"(" + binlogHeartbeatInterval + " * " + binlogHeartbeatIntervalAllowance + ")");
+			}
```

Comment:
have you managed to trigger this codepath (and the ensuing reconnection logic) locally?  Like by say, turning on maxwell's expectation of heartbeats but turning off the actual heartbeat mechanism? 

I don't necessarily expect you to write an integration test here, probably too hard, but I'd like to know that you at least reproduced a synthetic run of the code.

Reply:
Yeah, I'll hack up something locally once we're happy with the PR and make sure it hangs together 👍